src/main/haskell/kore/src/Kore/Step/Function/Matcher.hs

```diff
@@ -0,0 +1,431 @@
+        , SortedVariable variable
+        )
+    => MetadataTools level StepperAttributes
+    -> Map.Map (variable level) (variable level)
+    -> PureMLPattern level variable
+    -> PureMLPattern level variable
+    -> Maybe (Counter (PredicateSubstitution level variable))
+match tools quantifiedVariables first second =
+    firstJust
+        [ matchEqualHeadPatterns tools quantifiedVariables first second
+        , matchVariableFunction tools quantifiedVariables first second
+        , matchNonVarToPattern tools first second
+        ]
+  where
+    firstJust :: [Maybe a] -> Maybe a
+    firstJust = listToMaybe . catMaybes
+
+matchEqualHeadPatterns
+    :: ( Show (variable level)
+       , SortedVariable variable
+       , MetaOrObject level
+       , Ord (variable level)
+       , Ord (variable Meta)
+       , Ord (variable Object)
+       , Show (variable Meta)
+       , Show (variable Object)
+       , FreshVariable variable
+       , Hashable variable
+       )
+    => MetadataTools level StepperAttributes
+    -> Map.Map (variable level) (variable level)
+    -> PureMLPattern level variable
+    -> PureMLPattern level variable
+    -> Maybe (Counter (PredicateSubstitution level variable))
+matchEqualHeadPatterns tools quantifiedVariables first second =
+    case first of
+        (And_ _ firstFirst firstSecond) ->
+            case second of
+                (And_ _ secondFirst secondSecond) ->
+                    matchJoin
+                        tools
+                        quantifiedVariables
+                        [ (firstFirst, secondFirst)
+                        , (firstSecond, secondSecond)
+                        ]
+                _ -> Nothing
+        (App_ firstHead firstChildren) ->
+            case second of
+                (App_ secondHead secondChildren) ->
+                    if firstHead == secondHead
+                    then matchJoin
+                        tools
+                        quantifiedVariables
+                        (zip firstChildren secondChildren)
+                    else Nothing
+                _ -> Nothing
+        (Bottom_ _) ->
+            case second of
+                (Bottom_ _) -> Just $ return topPredicateSubstitution
+                _ -> Nothing
+        (Ceil_ _ _ firstChild) ->
+            case second of
+                (Ceil_ _ _ secondChild) ->
+                    match tools quantifiedVariables firstChild secondChild
+                _ -> Nothing
+        (CharLiteral_ _) ->
+            case second of
+                (CharLiteral_ _) ->
+                    if first == second
+                    then Just $ return topPredicateSubstitution
+                    else Nothing
+                _ -> Nothing
+        (DV_ _ _) ->
+            case second of
+                (DV_ _ _) ->
+                    if first == second
+                    then Just $ return topPredicateSubstitution
+                    else Nothing
+                _ -> Nothing
```

Comment:
In fact, that comment applies to each of these. `first == second` can only be true if they have the same head constructor, so the `case second of` is always redundant.